File: ios/Classes/OfflineRegionDefinition.swift
//
//  OfflineRegionDefinition.swift
//  mapbox_maps_flutter
//
//  Created by Margarita Zarubina on 19.12.2023.
//

import Foundation
import MapboxMaps

class OfflineRegionDefinition {
    let coordinates: Array<Array<Double>>
    let geometry: Geometry
    let mapStyleUrl: StyleURI
    let minZoom: Double
    let maxZoom: Double
    let radius: Double
    let metadata: [String : String]?
    let id: String
    
    init(coordinates: Array<Array<Double>>, mapStyleUrl: StyleURI, minZoom: Double, maxZoom: Double, metadata: [String : String]?, geometry: Geometry, id: String, radius: Double) {
        self.coordinates = coordinates
        self.geometry = geometry
        self.mapStyleUrl = mapStyleUrl
        self.minZoom = minZoom
        self.maxZoom = maxZoom
        self.metadata = metadata
        self.id = id
        self.radius = radius
    }
    
    static func fromDictionary(_ jsonDict: [String: Any]) -> OfflineRegionDefinition? {
        let radius = (jsonDict["radius"] as? Double) ?? 20000 // in m,
        
        guard let coordinates = jsonDict["coordinates"] as? Array<Array<Double>>,
              let mapStyleString = jsonDict["mapStyleUrl"] as? String,
              let minZoom = jsonDict["minZoom"] as? Double,
              let maxZoom = jsonDict["maxZoom"] as? Double,
              let id = jsonDict["id"] as? String,
              let mode = jsonDict["geometry"] as? String,
              let calculatedGeometry = Converters.geometryConverter(value: mode, coordinates: coordinates, radius: radius),
              let mapStyleUrl = StyleURI(rawValue: mapStyleString)
        else { return nil }
        return OfflineRegionDefinition(
            coordinates: coordinates,
            mapStyleUrl: mapStyleUrl,
            minZoom: minZoom,
            maxZoom: maxZoom,
            metadata: jsonDict["metadata"] as? [String: String],
            geometry: calculatedGeometry,
            id: id,
            radius: radius
        )
    }
}

class OfflineStyleDefinition {
    let mapStyleUrl: StyleURI
    let mode: GlyphsRasterizationMode
    let metadata: [String : String]?
    
    init(mode: GlyphsRasterizationMode?, mapStyleUrl: StyleURI,metadata : [String : String]?) {
        self.mode = mode ?? .ideographsRasterizedLocally
        self.mapStyleUrl = mapStyleUrl
        self.metadata = metadata
    }
    
    static func fromDictionary(_ jsonDict: [String: Any]) -> OfflineStyleDefinition? {
        guard let mode = jsonDict["mode"] as? String,
              let mapStyleString = jsonDict["mapStyleUrl"] as? String,
              let mapStyleUrl = StyleURI(rawValue: mapStyleString)
        else { return nil }
        return OfflineStyleDefinition(
            mode: Converters.glyphConverter(value: mode),
            mapStyleUrl: mapStyleUrl,
            metadata: jsonDict["metadata"] as? [String: String]
        )
    }
    
    
}

class Converters {
    static func glyphConverter(value: String) -> GlyphsRasterizationMode {
        switch(value){
        case "noGlyphsRasterizedLocally": return .noGlyphsRasterizedLocally
        case "allGlyphsRasterizedLocally": return .allGlyphsRasterizedLocally
        default: return .ideographsRasterizedLocally
        }
    }
    
    static func geometryConverter(value: String, coordinates: Array<Array<Double>>, radius: Double) -> Geometry? {
        switch(value){
        case "point": return .point(Point(LocationCoordinate2D(latitude: coordinates[0][0], longitude: coordinates[0][1])))
        case "lineString": return .lineString(LineString([LocationCoordinate2D(latitude: coordinates[0][0], longitude: coordinates[0][1]), LocationCoordinate2D(latitude: coordinates[1][0], longitude: coordinates[1][1])]))
        case "multiPoint": return .multiPoint(MultiPoint(coordinates.map { (coord) -> LocationCoordinate2D in
            return LocationCoordinate2D(latitude: coord[0], longitude: coord[1])
        }))
        case "multiPolygon":
            return .multiPolygon(MultiPolygon(coordinates.map { (coord) -> Polygon in
                return Polygon(center: CLLocationCoordinate2D(latitude: coord[0], longitude: coord[1]), radius: radius, vertices: 4)
            }))
        case "polygon": do {
            let center = Polygon(coordinates.map { (coord) -> [CLLocationCoordinate2D] in
                return [CLLocationCoordinate2D(latitude: coord[0], longitude: coord[1])]
                
            }).center
            return .polygon(Polygon(center: center ?? LocationCoordinate2D(latitude: coordinates[0][0], longitude: coordinates[0][1]), radius: radius, vertices: 4))
        }
            
        default: return nil
        }
    }
}

File: android/src/main/kotlin/com/mapbox/maps/mapbox_maps/offline/OfflineRegionDefinition.kt
package com.mapbox.maps.mapbox_maps.offline

import com.mapbox.geojson.*
import com.mapbox.maps.GlyphsRasterizationMode
import com.mapbox.turf.TurfConstants
import com.mapbox.turf.TurfMeasurement
import com.mapbox.turf.TurfTransformation
import java.lang.StrictMath.abs

class OfflineRegionDefinition(
  val geometry: Geometry,
  val mapStyleUrl: String,
  val minZoom: Double,
  val maxZoom: Double,
  val radius: Double,
  val metadata: Map<String, String>?,
  val id: String
) {
  companion object {
    fun fromDictionary(jsonDict: Map<String, Any>): OfflineRegionDefinition? {
      val radius = (jsonDict["radius"] as? Double) ?: 20000.0 // in m,
      val coordinates = jsonDict["coordinates"] as? List<List<Double>>? ?: return null
      val mapStyleString = jsonDict["mapStyleUrl"] as? String ?: return null
      val minZoom = jsonDict["minZoom"] as? Double ?: return null
      val maxZoom = jsonDict["maxZoom"] as? Double ?: return null
      val id = jsonDict["id"] as? String ?: return null
      val mode = jsonDict["geometry"] as? String ?: return null
      val calculatedGeometry = Converters.geometryConverter(mode, coordinates, radius) ?: return null

      return OfflineRegionDefinition(
        calculatedGeometry,
        mapStyleString,
        minZoom,
        maxZoom,
        radius,
        jsonDict["metadata"] as? Map<String, String>,
        id,
      )
    }
  }
}

class OfflineStyleDefinition(
  val mapStyleUrl: String,
  val mode: GlyphsRasterizationMode,
  val metadata: Map<String, String>?
) {

  companion object {
    fun fromDictionary(jsonDict: Map<String, Any>): OfflineStyleDefinition? {
      val mode = jsonDict["mode"] as? String ?: return null
      val mapStyleString = jsonDict["mapStyleUrl"] as? String ?: return null

      return OfflineStyleDefinition(
        mapStyleString,
        Converters.glyphConverter(mode),
        jsonDict["metadata"] as? Map<String, String>
      )
    }
  }
}

object Converters {
  fun glyphConverter(value: String): GlyphsRasterizationMode {
    return when (value) {
      "noGlyphsRasterizedLocally" -> GlyphsRasterizationMode.NO_GLYPHS_RASTERIZED_LOCALLY
      "allGlyphsRasterizedLocally" -> GlyphsRasterizationMode.ALL_GLYPHS_RASTERIZED_LOCALLY
      else -> GlyphsRasterizationMode.IDEOGRAPHS_RASTERIZED_LOCALLY
    }
  }

  fun geometryConverter(value: String, coordinates: List<List<Double>>, radius: Double): Geometry? {
    return when (value) {
      "point" -> Point.fromLngLat(coordinates[0][0], coordinates[0][1])
      "lineString" -> LineString.fromLngLats(coordinates.map { Point.fromLngLat(it[0], it[1]) })
      "multiPolygon" -> MultiPolygon.fromPolygons(coordinates.map {
        TurfTransformation.circle(Point.fromLngLat(it[0], it[1]), radius, 4, TurfConstants.UNIT_METERS)
      })
      "polygon" -> {
        val polygon = Polygon.fromLngLats(listOf(coordinates.map { Point.fromLngLat(it[0], it[1]) }))
        val bbox = TurfMeasurement.bbox(polygon)
        val center = Point.fromLngLat(abs(bbox[2] - bbox[0]) / 2, abs(bbox[3] - bbox[1]) / 2)
        TurfTransformation.circle(center, radius, 4, TurfConstants.UNIT_METERS)
      }
      else -> null
    }
  }
}